lfs/gitscanner_tree.go

```diff
@@ -130,76 +125,114 @@ func lsTreeBlobs(ref string, filter *filepathfilter.Filter) (*TreeBlobChannelWra
-}
+	pointers := make(map[string]*WrappedPointer)
 
-func (s *lsTreeScanner) TreeBlob() *TreeBlob {
-	return s.tree
-}
+	paths := make([]git.AttributePath, 0)
+	processor := gitattr.NewMacroProcessor()
 
-func (s *lsTreeScanner) Err() error {
-	return nil
-}
+	hasNext := true
+	for t := range treeblobs.Results {
+		if path.Base(t.Filename) == ".gitattributes" {
+			hasNext = oscanner.Scan(t.Oid)
 
-func (s *lsTreeScanner) Scan() bool {
-	t, hasNext := s.next()
-	s.tree = t
-	return hasNext
-}
+			if rdr := oscanner.Contents(); rdr != nil {
+				paths = append(paths, git.AttrPathsFromReader(
+					processor,
+					t.Filename,
+					"",
+					rdr,
+					t.Filename == ".gitattributes", // Read macros from the top-level attributes
+				)...)
+			}
 
-func (s *lsTreeScanner) next() (*TreeBlob, bool) {
-	hasNext := s.s.Scan()
-	line := s.s.Text()
-	parts := strings.SplitN(line, "\t", 2)
-	if len(parts) < 2 {
-		return nil, hasNext
-	}
+			if err := oscanner.Err(); err != nil {
+				return nil, nil, err
+			}
+		} else if t.Size < blobSizeCutoff {
+			hasNext = pscanner.Scan(t.Oid)
+
+			// It's intentional that we insert nil for
+			// non-pointers; we want to keep track of them
+			// as well as pointers.
+			p := pscanner.Pointer()
+			if p != nil {
+				p.Name = t.Filename
+			}
+			pointers[t.Filename] = p
 
-	attrs := strings.SplitN(parts[0], " ", 4)
-	if len(attrs) < 4 {
-		return nil, hasNext
+			if err := pscanner.Err(); err != nil {
+				return nil, nil, err
+			}
+		} else {
+			pointers[t.Filename] = nil
+		}
+
+		if !hasNext {
+			break
+		}
 	}
 
-	if attrs[1] != "blob" {
-		return nil, hasNext
+	// If the scanner quit early, we may still have treeblobs to
+	// read, so waiting for it to close will cause a deadlock.
+	if hasNext {
+		// Deal with nested error from incoming treeblobs
+		err := treeblobs.Wait()
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
-	sz, err := strconv.ParseInt(strings.TrimSpace(attrs[3]), 10, 64)
-	if err != nil {
-		return nil, hasNext
+	if err = pscanner.Close(); err != nil {
+		return nil, nil, err
+	}
+	if err = oscanner.Close(); err != nil {
+		return nil, nil, err
 	}
 
-	if sz < blobSizeCutoff {
-		sha1 := attrs[2]
-		filename := parts[1]
-		return &TreeBlob{Sha1: sha1, Filename: filename}, hasNext
+	patterns := make([]filepathfilter.Pattern, 0, len(paths))
+	for _, path := range paths {
+		// Convert all separators to `/` before creating a pattern to
+		// avoid characters being escaped in situations like `subtree\*.md`
+		patterns = append(patterns, filepathfilter.NewPattern(filepath.ToSlash(path.Path), filepathfilter.Strict(true)))
 	}
-	return nil, hasNext
+
+	return pointers, filepathfilter.NewFromPatterns(patterns, nil), nil
 }
 
-func scanNullLines(data []byte, atEOF bool) (advance int, token []byte, err error) {
-	if atEOF && len(data) == 0 {
-		return 0, nil, nil
+func runScanTreeForInvalidPointers(cb GitScannerFoundPointer, tree string, gitEnv, osEnv config.Environment) error {
```

Comment:
The name of this function is currently `runScanTreeForInvalidPointers()`, but if I understand the context in which it's run by `scanRefsByTree()`, and the way the `cb GitScannerFoundPointer` callback is used for both valid and invalid pointers, it seemed to me this might be renamed to just `runScanTreeForPointers()`.  Does that make sense?  (If so, I guess `catFileBatchTreeForInvalidPointers()` above might also be renamed?)

Reply:
Renamed.